File: src/test/java/net/slipp/domain/qna/QuestionBuilder.java
package net.slipp.domain.qna;

import java.util.List;
import java.util.Set;

import net.slipp.domain.tag.Tag;
import net.slipp.domain.user.SocialUser;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;

public class QuestionBuilder {
	private Long questionId;
	private SocialUser writer;
	private String title;
	private String contents;
	private Set<Tag> tags = Sets.newHashSet();
	private List<Answer> answers = Lists.newArrayList();

	private QuestionBuilder() {}

	private QuestionBuilder(Long questionId) {
		this.questionId = questionId;
	}
	
	public static QuestionBuilder aQuestion() {
		return new QuestionBuilder();
	}

	public static QuestionBuilder aQuestion(Long questionId) {
		return new QuestionBuilder(questionId);
	}
	
	public QuestionBuilder withWriter(SocialUser writer) {
		this.writer = writer;
		return this;
	}
	
	public QuestionBuilder withTitle(String title) {
		this.title = title;
		return this;
	}
	
	public QuestionBuilder withContents(String contents) {
		this.contents = contents;
		return this;
	}
	
	public QuestionBuilder withTag(Tag tag) {
		tags.add(tag);
		return this;
	}
	
	public QuestionBuilder withAnswer(Answer answer) {
		answers.add(answer);
		return this;
	}
	
	public Question build() {
		Question question = new Question(questionId, writer, title, contents, tags) {
			public List<Answer> getAnswers() {
				return answers;
			}
		};
		for (Answer answer : answers) {
			question.newAnswered(answer);
		}
		
		return question;
	}
}

File: src/test/scala/net/slipp/domain/qna/QuestionBuilder.scala
package net.slipp.domain.qna

import java.lang.Long
import java.util.List

import com.google.common.collect.{Lists, Sets}
import net.slipp.domain.tag.Tag
import net.slipp.domain.user.SocialUser

object QuestionBuilder {
  def aQuestion: QuestionBuilder = {
    return new QuestionBuilder(null)
  }

  def aQuestion(questionId: Long): QuestionBuilder = {
    return new QuestionBuilder(questionId)
  }
}

class QuestionBuilder(questionId: Long) {
  private var writer: SocialUser = null
  private var title: String = null
  private var contents: String = null
  private var tags = Sets.newHashSet[Tag]
  private var answers = Lists.newArrayList[Answer]

  def withWriter(writer: SocialUser): QuestionBuilder = {
    this.writer = writer
    return this
  }

  def withTitle(title: String): QuestionBuilder = {
    this.title = title
    return this
  }

  def withContents(contents: String): QuestionBuilder = {
    this.contents = contents
    return this
  }

  def withTag(tag: Tag): QuestionBuilder = {
    tags.add(tag)
    return this
  }

  def withAnswer(answer: Answer): QuestionBuilder = {
    answers.add(answer)
    return this
  }

  def build: Question = {
    val question: Question = new Question(questionId, writer, title, contents, tags) {
      override def getAnswers: List[Answer] = {
        return answers
      }
    }
    import scala.collection.JavaConversions._
    for (answer <- answers) {
      question.newAnswered(answer)
    }
    return question
  }
}
